### PR TITLE
add new es client

### PR DESCRIPTION
**Pull request by cangfengzhs**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [ ] enhancement

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [x] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+  body["query"] = folly::dynamic::object();
+  body["query"]["fuzzy"] = folly::dynamic::object();
+  body["query"]["fuzzy"]["text"] = folly::dynamic::object("value", pattern)("fuzziness", fuzziness);
+  return ESAdapter::query(index, body);
+}
+
+// StatusOr<ESQueryResult> ESAdapter::term(const std::string& index,const std::vector<std::string>&
+// words){
+//   folly::dynamic query = folly::dynamic::object("query",folly::dynam)
+// }
```
Comment (panda-sheep) [inline src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp:218]: Can this commented code be removed?

File: src/common/plugin/fulltext/elasticsearch/ESClient.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+// StatusOr<folly::dynamic> ESClient::put(const std::string& index, const folly::dynamic& document)
+// {
+//   std::string url = fmt::format("http://{}/{}/_doc", address_, index);
+//   auto resp = httpClient_.post(url, {"Content-Type: application/json"}, folly::toJson(document));
+//   if (resp.curlCode != 0) {
+//     return Status::Error(fmt::format("curl error({}):{}", resp.curlCode, resp.curlMessage));
+//   }
+//   return folly::parseJson(resp.body);
+// }
```
Comment (panda-sheep) [inline src/common/plugin/fulltext/elasticsearch/ESClient.cpp:107]: ditto

File: src/common/plugin/fulltext/test/CMakeLists.txt [hunk 3]
```diff
@@ -1,21 +1,37 @@
...
+#         $<TARGET_OBJECTS:base_obj>
+#         $<TARGET_OBJECTS:network_obj>
+#         $<TARGET_OBJECTS:fs_obj>
+#         $<TARGET_OBJECTS:process_obj>
+#         $<TARGET_OBJECTS:es_adapter_obj>
+#     LIBRARIES
+#         ${THRIFT_LIBRARIES}
+#         ${PROXYGEN_LIBRARIES}
+#         gtest
+# )
```
Comment (panda-sheep) [inline src/common/plugin/fulltext/test/CMakeLists.txt:20]: ditto

File: src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp (hunk 1 above)
Comment (cangfengzhs) [reply src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp:218]: This part of the code will be used, but we haven't fully figured out how to implement it. Uncomment and test them in the next PR

File: src/common/plugin/fulltext/elasticsearch/ESClient.cpp (hunk 2 above)
Comment (cangfengzhs) [reply src/common/plugin/fulltext/elasticsearch/ESClient.cpp]: done

File: src/common/plugin/fulltext/test/CMakeLists.txt (hunk 3 above)
Comment (cangfengzhs) [reply src/common/plugin/fulltext/test/CMakeLists.txt]: done

**Review by panda-sheep — APPROVED** [review]
lgtm

File: src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    hex[i * 2] = 'A' + mdStr[i] / 16;
+    hex[i * 2 + 1] = 'A' + mdStr[i] % 16;
+  }
+  return std::string(hex, 64);
+}
+
+ESClient& ESAdapter::randomClient() {
+  static thread_local std::default_random_engine engine;
+  static thread_local std::uniform_int_distribution<size_t> d(0, clients_.size() - 1);
+  return clients_[d(engine)];
```
Comment (critical27) [inline src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp:278]: I don't quite remember, do we already support sign in multiple es client?

**Review by critical27 — COMMENTED** [review]
Good job

File: src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp (hunk 4 above)
Comment (cangfengzhs) [reply src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp:278]: I did not verify it, but it is supported in grammar and implementation.

Comment (critical27) [reply src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp:278]: 👌

**Review by critical27 — APPROVED** [review]

**Pull request merged**